rustls/src/conn/kernel.rs

```diff
@@ -0,0 +1,273 @@
+    /// connection. Attempting to do so on a non-TLS 1.3 connection will result
+    /// in an error.
+    pub fn update_rx_secret(&mut self) -> Result<(u64, ConnectionTrafficSecrets), Error> {
+        // The sequence number always starts at 0 after a key update.
+        self.state
+            .update_secrets(Direction::Receive)
+            .map(|secret| (0, secret))
+    }
+}
+
+impl KernelConnection<ClientConnectionData> {
+    /// Handle a `new_session_ticket` message from the peer.
+    ///
+    /// This will register the session ticket within with rustls so that it can
+    /// be used to establish future TLS connections.
+    ///
+    /// # Getting the right payload
+    ///
+    /// This method expects to be passed the inner payload of the handshake
+    /// message. This means that you will need to parse the header of the
+    /// handshake message in order to determine the correct payload to pass in.
+    /// The message format is described in [RFC 8446 section 4][0]. `payload`
+    /// should not include the `msg_type` or `length` fields.
+    ///
+    /// Code to parse out the payload should look something like this
+    /// ```no_run
+    /// use rustls::{ContentType, HandshakeType};
+    /// use rustls::kernel::KernelConnection;
+    /// use rustls::client::ClientConnectionData;
+    ///
+    /// # fn doctest(conn: &mut KernelConnection<ClientConnectionData>, typ: ContentType, message: &[u8]) -> Result<(), rustls::Error> {
+    /// let conn: &mut KernelConnection<ClientConnectionData> = // ...
+    /// #   conn;
+    /// let typ: ContentType = // ...
+    /// #   typ;
+    /// let mut message: &[u8] = // ...
+    /// #   message;
+    ///
+    /// // Processing for other messages not included in this example
+    /// assert_eq!(typ, ContentType::Handshake);
+    ///
+    /// // There may be multiple handshake payloads within a single handshake message.
+    /// while !message.is_empty() {
+    ///     let (typ, len, rest) = match message {
+    ///         &[typ, a, b, c, ref rest @ ..] => (
+    ///             HandshakeType::from(typ),
+    ///             u32::from_be_bytes([0, a, b, c]) as usize,
+    ///             rest
+    ///         ),
+    ///         _ => panic!("error handling not included in this example")
+    ///     };
+    ///
+    ///     // Processing for other messages not included in this example.
+    ///     assert_eq!(typ, HandshakeType::NewSessionTicket);
+    ///     assert!(rest.len() >= len, "invalid handshake message");
+    ///
+    ///     let (payload, rest) = rest.split_at(len);
+    ///     message = rest;
+    ///
+    ///     conn.handle_new_session_ticket(payload)?;
+    /// }
+    /// # Ok(())
+    /// # }
+    /// ```
+    ///
+    /// # Errors
+    /// This method will return an error if:
+    /// - This connection is not a TLS 1.3 connection (in TLS 1.2 session tickets
+    ///   are sent as part of the handshake).
+    /// - This connection is a server connection.
```

Comment:
nit: since this is defined on `KernelConnection<ClientConnectionData>` I don't think `KernelConnection<ServerConnectionData>` gets to return an error.

Reply:
Good catch! That was left over from when I made `KernelConnection` generic. I must have missed it when making changes.